Implement call graph pruning feature

bingtimren opened the pull request:
Generating large call graph can be slow and consumes large memory.

This feature allows pruning the call graph with a bound function, so to generate an incomplete call graph according to the analysis algorithm user's specific needs - with potentially less memory & CPU usage.

brenhub24 committed:
Implement call graph pruning feature

swissiety commented on sootup.callgraph/src/main/java/sootup/callgraph/AbstractCallGraphAlgorithm.java:
```diff
@@ -232,7 +235,7 @@ protected void resolveAllCallsFromSourceMethod(
         .forEach(
             stmt ->
-                resolveCall(sourceMethod, stmt)
+                 (boundFunction == null || Boolean.TRUE.equals(boundFunction.apply(sourceMethod, stmt)) ? resolveCall(sourceMethod, stmt) : Stream.<MethodSignature> empty())
```
1) instead of null lets make use of a method that just returns always true
2) lets use BiPredicate as type to reduce unnecessary AutoBoxing and simplify the condition check  to just the method call

bingtimren replied:
will make adjustment as suggested

swissiety commented on sootup.callgraph/src/main/java/sootup/callgraph/AbstractCallGraphAlgorithm.java:
```diff
@@ -586,4 +589,25 @@ public static Optional<SootMethod> findConcreteMethod(
+   * cost.
+   * 
+   * @param boundFunction */
+  public void setBoundFunction(BiFunction<SootMethod, InvokableStmt, Boolean> boundFunction)
+  {
+      this.boundFunction = boundFunction;
+  }
```
lets set the function in the constructor only

bingtimren replied:
hmm.... but this would force all sub-classes implement two constructors - one with a bound function, one without (so to use the always-true function)..... instead of having this feature automatically

let's keep it this way, or change setter to a "with" method (fluent API)?

bingtimren replied:
if you are concerned bound function is set after constructCompleteCallGraph() is called, I can add a checker and throw an exception if user do that. thought?

swissiety replied:
lets implement it as a method directly without BiPredicate etc. - that way someone can override it the same way its done with a linkedhashmap to transform it to a LRUcache via overriding removeoldestentry() - e.g. see https://ashcode.medium.com/lru-cache-using-linkedhashmap-java-85e78ee1b976

bingtimren replied:
I would still recommend keeping it this way. This is more a setting of the analyser than an extension to it. In general I would try to avoid overriding a non-abstract method, it's kind of breaking encapsulation. In general composition is preferred than inheritance, see https://en.wikipedia.org/wiki/Composition_over_inheritance

JonasKlauke replied:
I think if you configure it in the constructor, the developer might be more aware of the feature.

swissiety reviewed (commented):
thanks for that contribution! I will merge it with just some small adjustments.

brenhub24 committed:
use of a method that just returns always true instead of null
lets use BiPredicate as type to reduce unnecessary AutoBoxing and
simplify the condition check to just the method call

bingtimren committed:
Merge branch 'develop' into feature/branchPruning

moebarni committed:
Merge branch 'develop' into feature/branchPruning

# Conflicts:
#	sootup.callgraph/src/main/java/sootup/callgraph/AbstractCallGraphAlgorithm.java
#	sootup.tests/src/test/java/sootup/tests/CallGraphTest.java

moebarni committed:
wip (aligned with current develop)

moebarni committed:
wip

moebarni committed:
Merge branch 'develop' into feature/branchPruning

moebarni committed:
fixed formalities

moebarni committed:
Merge branch 'develop' into feature/branchPruning

moebarni committed:
deleted setBoundFunction; second const; with BiPredicate

moebarni committed:
deleted boundFunction, getBoundFunction; one const only; old tests

moebarni committed:
wip (AsmJavaClassProvider - Class names do not match.)

moebarni committed:
missing package fixed; new tests; deleted old tests

JonasKlauke commented on sootup.callgraph/src/test/java/sootup/callgraph/RapidTypeAnalysisPruningAlgorithmTest.java:
```diff
@@ -0,0 +1,60 @@
+import sootup.java.core.views.JavaView;
+
+public class RapidTypeAnalysisPruningAlgorithmTest
+    extends CallGraphTestBase<RapidTypeAnalysisPruningAlgorithm> {
```
Solution without extends.
Maybe use Util Class for views and so on

moebarni committed:
Merge branch 'develop' into feature/branchPruning

# Conflicts:
#	sootup.callgraph/src/test/java/sootup/callgraph/CallGraphTestBase.java

moebarni committed:
CHA-Test; CallGraphTestMethods as superclass of CallGraphTestBase

JonasKlauke committed:
Merge branch 'develop' into feature/branchPruning
